magma/digital.py

```diff
@@ -0,0 +1,198 @@
+        if value is not None:
+            if isinstance(value, (bool, IntegerTypes)):
+                return VCC if value else GND
+        result = super().__call__(*args, **kwargs)
+        if value is not None:
+            assert isinstance(value, Digital), type(value)
+            result(value)
+        return result
+
+    def __repr__(cls):
+        return str(cls)
+
+    def __str__(cls):
+        name = getattr(cls, "orig_name", cls.__name__)
+        if cls.is_input():
+            name = f"In({name})"
+        elif cls.is_output():
+            name = f"Out({name})"
+        elif cls.is_inout():
+            name = f"InOut({name})"
+        return name
+
+    @property
+    def undirected_t(cls) -> 'DigitalMeta':
+        t = cls._info_[0]
+        if t is not None:
+            return t
+        else:
+            raise AttributeError('type {} has no undirected_t'.format(cls))
+
+    @property
+    def direction(cls) -> int:
+        return cls._info_[1]
+
+    @property
+    def is_directed(cls) -> bool:
+        return cls.direction is not None
+
+    def __len__(cls):
+        return 1
+
+    def qualify(cls, direction):
+        return cls[direction]
+
+    def flip(cls):
+        return cls.qualify(Direction.Flip)
+
+    def __eq__(cls, rhs):
+        return cls is rhs
+
+    __hash__ = type.__hash__
+
+
+class Digital(Type, metaclass=DigitalMeta):
+    def __init__(self, value=None, name=None):
+        super().__init__(name=name)
+        self.port = Port(self)
+
+    @classmethod
+    def is_oriented(cls, direction):
+        return cls.direction == direction
+
+    def __call__(self, output):
+        return self.wire(output, get_callee_frame_info())
+
+    @debug_wire
+    def wire(self, o, debug_info):
+        i = self
+        # promote integer types to LOW/HIGH
+        if isinstance(o, IntegerTypes):
+            o = HIGH if o else LOW
+
+        if not isinstance(o, Digital):
+            report_wiring_error(f'Cannot wire {i.debug_name} (type={type(i)}) '
+                                f'to {o} (type={type(o)}) because '
+                                f'{o.debug_name} is not a Digital', debug_info)
+            return
+
+        i.port.wire(o.port, debug_info)
+        i.debug_info = debug_info
+        o.debug_info = debug_info
+
+    def wired(self):
+        return self.port.wired()
+
+    # return the input or output Bit connected to this Bit
+    def trace(self):
+        t = self.port.trace()
+        if t:
+            t = t.bit
+        return t
+
+    # return the output Bit connected to this input Bit
+    def value(self):
+        t = self.port.value()
+        if t:
+            t = t.bit
+        return t
+
+    def driven(self):
+        return self.port.driven()
+
+    def flatten(self):
+        return [self]
+
+    def const(self):
+        return self is VCC or self is GND
+
+    def unwire(i, o):
+        i.port.unwire(o.port)
+
+    @classmethod
+    def flat_length(cls):
+        return 1
+
```

Comment:
probably loam was supposed to be updated to sync with magma hwtypes2, but in the meantime, these two methods were required to build with loam. 

    def getinst(self):
        t = self.trace()
        return t.name.inst if t is not None else None

    def getgpio(self):
        return self.getinst()

- https://github.com/phanrahan/loam/search?q=getinst&unscoped_q=getinst
- https://github.com/phanrahan/loam/search?q=getgpio&unscoped_q=getgpio

Reply:
Thanks, I've added these in for compatibility, although we should revisit this interface as it seems pretty loam specific so perhaps they should not be part of the magma core.